enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    ...
+```
+
+When the new `pinnedImages` field is added or changed the machine config
+operator will need to pre-load those images (with the equivalent of `crictl
+pull`), create or update the corresponding
+`/etc/crio/crio.conf.d/pin-upgrade.conf` file and ask CRI-O reload its
+configuration (with the equivalent of `systemctl reload crio.service`).
+
+If the `spec.containerRuntimeConfig.imagesDirectory` field is used then it
+should contain a dump of a registry server. The machine config operator will
+first check if there is enough space available in the node to copy the images
+to the `/var/lib/containers/storage` directory. Images in the dump of a
+registry server are compressed, but in `/var/lib/containers/storage` they are
+not, so the machine config operator will check that there is at least twice the
+space space used by the registry server dump. If there is not enough space it
+will be reported as an error conditions in the `ContainerRuntimeConfig` object.
+
+If there is enough space then the machine config operator will start a
+temporary embedded image registry server in each node of the cluster, listening
+in a randomly selected local port, and using a self signed certificate, to serve
+the images from the `/var/lib/upgrade/4.13.17-x86_64` directory. It will then
+configure the node to trust the self signed certificate and configure CRI-O to
+use that temporary image registry server as a mirror for the pinned images,
+creating a `/etc/containers/registries.conf.d/pin-upgrade.conf` file with a
+content similar to this:
+
+```toml
+[[registry]]
+prefix = "quay.io/openshift-release-dev/ocp-release"
+location = "quay.io/openshift-release-dev/ocp-release"
+
+[[registry.mirror]]
+location = "localhost:12345/openshift-release-dev/ocp-release"
+
+[[registry]]
+prefix = "quay.io/openshift-release-dev/ocp-v4.0-art-dev"
+location = "quay.io/openshift-release-dev/ocp-v4.0-art-dev"
+
+[[registry.mirror]]
+location = "localhost:12345/openshift-release-dev/ocp-v4.0-art-dev"
+```
+
+The machine config operator will copy the images to the
+`/var/lib/containers/storage` directory. To do that it will use the gRPC API of
+CRI-O to run the equivalent of `crictl pull` for each of the images. When that
+is completed the machine config operator will update the new
+`status.pinnedImages` field of the `ContainerRuntimeConfig` object:
+
+```yaml
+status:
+  pinnedImages:
+  - quay.io/openshift-release-dev/ocp-release@sha256:...
+  - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  ...
+```
+
+We explored several alternatives for the format of the bundle, and using a dump
+of a registry server to be the best one.
```

Comment:
Really?  There isn't "a" registry server, which specific registry server?  docker/distribution?

Did you investigate `oci` directories?

Reply:
Yes, I mean https://github.com/distribution/distribution, will update the comment.

I did try to use `oci` directories and archives. The problem I found is that the OCI format requires use of the `application/vnd.oci.image.manifest.v1+json` media type in manifests, but our images use `application/vnd.docker.distribution.manifest.v2+json`. That means that when the image is written in OCI format its manifest is different, and therefore the digest is different. The upgrade pulls images by digest, so these modified images don't work. There is a `--preserve-digest` option in `skopeo copy`, but I think it can't address this issue. To make this work we would have to change the digests in the release image and in all the other places where those digests are used.